File: ch5/Shell_Sort/algorithm.java
import java.util.Arrays;

class algorithm {
    public static <T extends Comparable<T>> void InsertionSubSort(T[] list, int k) {
        for (int i = k ; i < list.length; i++) {
            T current = list[i];
            int position = i - k;
            while ((position >= 0) && (current.compareTo(list[position]) < 0)) {
                list[position + k] = list[position];
                position = position - k;
            }
            list[position + k] = current;
        }
    }

    public static <T extends Comparable<T>> void ShellSort(T[] list, int[] k) {
        for (int i = 0; i < k.length ; i++) {
            InsertionSubSort(list, k[i]);
        }
    }

    public static void main(String args[]) {
        Integer arr[] = {12,11,13,1,5,6};
        int k[] = {6,5,4,3,2,1};
        ShellSort(arr, k);
        System.out.println("Sorted list: " + Arrays.toString(arr));
    }
}

File: ch5/Shell_Sort/algorithm.go
package main
import "fmt"

func InsertionSubSort(L []int, k int) {
	for i:= k; i < len(L); i++ {
		current := L[i]
		position := i - k
		for position >= 0 && current < L[position] {
			L[position + k] = L[position]
			position -= k
		}
		L[position + k] = current
	}
}

func ShellSort(L []int, k []int) {
	for i:=0 ; i < len(k) ; i++ {
		InsertionSubSort(L, k[i])
	}
}

func main() {
	List := []int{4,3,2,0,12,10,2,9,8,-1,-2,3,6,58,8,34,54,62,61,54,50}
	k := []int{6,5,4,3,2,1}
	ShellSort(List, k)
	fmt.Println(List)
}